### PR TITLE
Fire diagnostic source events from IHostBuilder.Build

### PR DESCRIPTION
**Pull request by davidfowl**
- We want to enable getting access to the `IHostBuilder` and `IHost` during the call to `IHostBuilder.Build` so that we can access the application's `IServiceProvider` from various tools (like the EntityFramework migrations tool or scaffolding). Usually, this is enabled by exposing a different entry point from the application (`CreateHostBuilder`) but this technique allows us to let the application execute normally and hook important events via a diagnostic source.
- Updated HostFactoryResolver and added new APIs for resolving the IHostBuilder from an assembly with an entry point. 
- Added tests

~TODO: I need to update the `HostFactoryResolver` to support this new pattern.~

~TODO: Add tests for various scenarios:~
- ~Exception thrown before IHostBuilder.Build is called where the exception propagates~
- ~Exception thrown before IHostBuilder.Build is called where the exception does not propagate~
- Crazy scenarios where multiple host builders are fired up at during the call to main (not sure this is important)

cc @halter73

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:262]: This is where we wire up to the new event to intercept calls to build.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:207]: I'm not sure if we should propagate the OperationCanceledException here or return null.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:158]: I'm not sure what a decent default should be. I had 30 seconds but that would make testing invalid tests kinda awful.

**Comment by martincostello** [comment]
Would this light-up the ability to use `WebApplicationFactory<T>` (and maybe top-level programs) with the new minimal hosting APIs?

**Comment by davidfowl** [comment]
Yes that's the intent:
- EF migrations
- WebApplicationFactory
- NSwag
- Scaffolding 

Etc

**Comment by martincostello** [comment]
Nice 👍 - I've been holding off in using it in a bunch of stuff as so much of our tests use WebApplicationFactory to do integration-level tests.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:134]: I know we have the tests in the HostFactoryResolver, but it might be a good idea to add some tests directly to Extensions.Hosting that ensure this fires correctly. To prevent breaking it in the future.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: It isn't the "The values being passed into Write" that are the issue, but instead the recursive properties of the values.

DiagnosticSource has an EventSource bridge that allows end-users to set up "filter specs" that specify which properties of the objects passed here should be written to the EventSource. If those properties aren't used in the app, they could be trimmed.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: I think that's true in general but in the use case for these events, I'm not worried:
- There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.
- In the 90% case, the relevant properties/methods are going to be used as part of application bootstrapping.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: > There won't be any trimming happening in development. You don't publish an application trimmed and then run tools in it to grab the IHost and IHostBuilder.

What about someone else who takes a dependency on this new DiagnosticSource?

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: > Of course, if we don't expect people to be using the EventSource bridge, that could be a "good enough" reason to suppress.

Yes this isn't a use case. There's nothing serializable on there that can be sent out of process.

> What about someone else who takes a dependency on this new DiagnosticSource?

Then we'll learn about the scenario then because it won't work for that scenario we don't understand.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:134]: Done

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:40]: 1. Does this need to be public?
2. If it does, it is a bit concerning that it has the same name as the method above, takes the same parameters, and returns the same type, but doesn't really do the same things. Can we rename it somehow to distinguish it from the existing method above?

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:158]:
```suggestion
            private static readonly TimeSpan s_waitTimeout = TimeSpan.FromSeconds(20);
```

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:164]:
```suggestion
                _configure = innerBuilder =>
```

or something similar. To make it easier to understand what this variable is.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:232]:
```suggestion
                    // Lets this propagate out of the call to GetAwaiter().GetResult()
```

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:227]: Maybe use a different message here - specifically call out that it timed out. That way we can tell the difference between the entrypoint returning gracefully vs. timing out.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/CreateHostBuilderInvalidSignature/Program.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/CreateHostBuilderInvalidSignature/Program.cs:5]: Are the changes to this file and `tests/CreateHostBuilderPatternTestSite/Program.cs` necessary?

**Review by eerhardt — APPROVED** [review]
Looks fine to me. Just some random nits.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/CreateHostBuilderInvalidSignature/Program.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/tests/CreateHostBuilderInvalidSignature/Program.cs:5]: Because I added Hosting as a dependency yes.

**Review by bricelam — APPROVED** [review]
LGTM. I really like this approach--it's the magic I've been looking for since the beginning of ASP.NET Core.

As discussed offline, we should try hard to avoid the compile-time dependency on `Microsoft.Extensions.Hosting`. We had a lot of pushback against this in the 1.1 days since it adds it to all app models (see https://github.com/dotnet/efcore/issues/7835 from a WPF user for example)

/cc @ajcvickers

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (halter73) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:235]: Nit: I think the when makes this clear enough

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs
Comment (halter73) [inline src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:152]: Any particular reason for resolving the IHost before writing this event? I get that it resolves a bunch of services. Is that important for any particular reason? Why not leave that to the listener?

Comment (halter73) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:152]: I guess the `_hostBuilt` flag being set would make it tricky to build in the subscriber once you get this far.

**Review by halter73 — APPROVED** [review]

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:227]: I want the timeout to be an implementation detail. I'm also thinking this could also return null.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:40]: I'll rename it.

File: src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs
Comment (noahfalk) [reply src/libraries/Microsoft.Extensions.Hosting/src/HostBuilder.cs:159]: I think anyone else who used the DiagnosticSource would either:
a) cast the received object to its correct type which would make the linker aware of the usage - everything works nicely here
b) private reflection - I trust this puts the onus on the new DiagnosticSource consumer to ensure the fields aren't removed

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (noahfalk) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:53]: Nit:
```suggestion
                // even have a chance of firing (they were added in .NET >= 6)
```

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (noahfalk) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:236]: Who do you expect to see the message? If the message is primarily shown to people who are trying to resolve the issue (or users will relay it to someone else who resolves the issue) then it might help to make it more descriptive.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (noahfalk) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:281]: The prize for dirty hackery this week : D

Is it important to guarantee that the thread really does stop? It would be easy enough for an app to throw a try/catch around the part of their code where this gets raised so that it never gets back to the entrypoint. Your tool code will be running in parallel with the user's unknown error handling app code. It doesn't seem obviously harmful to me but figured I mention it.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (eerhardt) [inline src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:43]: I don't see a test that uses `configureHostBuilder`. Do we need this parameter?

**Comment by davidfowl** [comment]
@eerhardt and @noahfalk I need to get this merged so I can react in several places down stream. I will react to your feedback in another PR so keep commenting.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:281]: Disposal is one of the things I'm worried about and why I think throwing to stop execution makes the most sense here.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:236]: I need to check the callers but its possible null is a better return than throwing.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:43]: Yes, we will need it for https://github.com/dotnet/aspnetcore/blob/main/src/Mvc/Mvc.Testing/src/WebApplicationFactory.cs

Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:43]: Ok, then it would be great if we had a test ensuring it doesn't break.

Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:43]: 2 things:
1. As of right now, ASP.NET Core will need to use it so it kinda *has* to work or tests will fail regardless 😄.
2. I agree with you and will add a test.

File: src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs
Comment (davidfowl) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:281]: > It would be easy enough for an app to throw a try/catch around the part of their code where this gets raised so that it never gets back to the entrypoint.

The thing that stops this from happening in the throw case is the fact that the application never gets access to the IHost instance. They can't call build again on it because double building throws. Even if they catch the exception and do something else, the IHost isntance that came out of Build is never observed by the application.

Comment (noahfalk) [reply src/libraries/Microsoft.Extensions.HostFactoryResolver/src/HostFactoryResolver.cs:281]: Yeah, I wasn't imagining they continue normally, more like they have some complex error handling code that will be running in parallel. 

> the application never gets access to the IHost instance

[Joking] What do you mean? This PR just added the official mechanism that lets everyone access IHost without the pesky inconvenience of needing Build() to return it to you ;p

**Pull request merged**
